torch/csrc/jit/codegen/cuda/runtime/block_sync_atomic.cu

```diff
@@ -0,0 +1,54 @@
+
+// Counter-based block synchronization. Only meant to be used for
+// debugging and validating synchronization. This should be replaced
+// with cuda::barrier::arrive_and_wait as that should be more robust.
+
+namespace block_sync {
+
+using CounterType = unsigned int;
+static constexpr CounterType COUNTER_TYPE_MAX = ~(CounterType)0;
+__shared__ CounterType sync_counter;
+
+__device__ void init() {
+  const unsigned int tid = threadIdx.x + threadIdx.y * blockDim.x +
+      threadIdx.z * blockDim.x * blockDim.y;
+  if (tid == 0) {
+    sync_counter = 0;
+  }
+  __syncthreads();
+}
+
+// Emulate __syncthreads() with a synchronization counter
+__device__ void sync() {
+  unsigned int backoff = 8;
+  const unsigned int backoff_max = 256;
+  const unsigned int num_threads = blockDim.x * blockDim.y * blockDim.z;
+
+  __threadfence_block();
+
+  // Use counter range only up to a limit so that the next val won't
+  // overflow.
+
+  const auto counter_max = (COUNTER_TYPE_MAX / num_threads) * num_threads;
+  const auto old = atomicInc(&sync_counter, counter_max - 1);
+
+  const auto next = (old / num_threads) * num_threads + num_threads;
+
+  // Does this guarantee an actual read from shmem?
```

Comment:
I don't know, why not just make sync_counter volatile directly?

Reply:
That's an option, but then need to cast `&sync_counter` to `unsigned int*` when calling `atomicInc` as it doesn't accept `volatile unsigned int*`. Either looks ok to me.

Reply:
Seems fine to me then, can you just clean up the commented code?

Reply:
Thanks. Deleted it.